programs/drift/src/instructions/lp_pool.rs

```diff
@@ -0,0 +1,1554 @@
+
+    if !is_oracle_valid_for_action(in_oracle_validity, Some(DriftAction::LpPoolSwap))? {
+        msg!(
+            "In oracle data for spot market {} is invalid for lp pool swap.",
+            in_spot_market.market_index,
+        );
+        return Err(ErrorCode::InvalidOracle.into());
+    }
+
+    if !is_oracle_valid_for_action(out_oracle_validity, Some(DriftAction::LpPoolSwap))? {
+        msg!(
+            "Out oracle data for spot market {} is invalid for lp pool swap.",
+            out_spot_market.market_index,
+        );
+        return Err(ErrorCode::InvalidOracle.into());
+    }
+
+    update_spot_market_cumulative_interest(&mut in_spot_market, Some(&in_oracle), now)?;
+    update_spot_market_cumulative_interest(&mut out_spot_market, Some(&out_oracle), now)?;
+
+    let in_target_weight = constituent_target_base.get_target_weight(
+        in_constituent.constituent_index,
+        &in_spot_market,
+        in_oracle.price,
+        lp_pool.last_aum,
+    )?;
+    let out_target_weight = constituent_target_base.get_target_weight(
+        out_constituent.constituent_index,
+        &out_spot_market,
+        out_oracle.price,
+        lp_pool.last_aum,
+    )?;
+
+    let (in_amount, out_amount, in_fee, out_fee) = lp_pool.get_swap_amount(
+        &in_oracle,
+        &out_oracle,
+        &in_constituent,
+        &out_constituent,
+        &in_spot_market,
+        &out_spot_market,
+        in_target_weight,
+        out_target_weight,
+        in_amount as u128,
+        constituent_correlations.get_correlation(
+            in_constituent.constituent_index,
+            out_constituent.constituent_index,
+        )?,
+    )?;
+    msg!(
+        "in_amount: {}, out_amount: {}, in_fee: {}, out_fee: {}",
+        in_amount,
+        out_amount,
+        in_fee,
+        out_fee
+    );
+    let out_amount_net_fees = if out_fee > 0 {
+        out_amount.safe_sub(out_fee.unsigned_abs())?
+    } else {
+        out_amount.safe_add(out_fee.unsigned_abs())?
+    };
+
+    validate!(
+        out_amount_net_fees.cast::<u64>()? >= min_out_amount,
+        ErrorCode::SlippageOutsideLimit,
+        format!(
+            "Slippage outside limit: out_amount_net_fees({}) < min_out_amount({})",
+            out_amount_net_fees, min_out_amount
+        )
+        .as_str()
+    )?;
+
+    validate!(
+        out_amount_net_fees.cast::<u64>()? <= out_constituent.token_balance,
+        ErrorCode::InsufficientConstituentTokenBalance,
+        format!(
+            "Insufficient out constituent balance: out_amount_net_fees({}) > out_constituent.token_balance({})",
+            out_amount_net_fees, out_constituent.token_balance
+        )
+        .as_str()
+    )?;
+
+    in_constituent.record_swap_fees(in_fee)?;
+    out_constituent.record_swap_fees(out_fee)?;
+
+    let in_swap_id = get_then_update_id!(in_constituent, next_swap_id);
+    let out_swap_id = get_then_update_id!(out_constituent, next_swap_id);
+
+    emit!(LPSwapRecord {
```

Comment:
should use the custom emit_event marco we have which saves CU and avoids heap allocaitons